src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibraryWithFSharp.cs

```diff
@@ -0,0 +1,264 @@
+                .Execute()
+                .Should()
+                .Fail();
+        }
+
+        [Fact]
+        public void Restore_succeeds_even_if_the_project_extension_is_for_a_different_language()
+        {
+            var testAsset = _testAssetsManager
+                .CopyTestAsset("AppWithLibraryFS")
+                .WithSource();
+
+            var libraryProjectDirectory = Path.Combine(testAsset.TestRoot, "TestLibrary");
+
+            var oldProjectFile = Path.Combine(libraryProjectDirectory, "TestLibrary.fsproj");
+            var newProjectFile = Path.Combine(libraryProjectDirectory, "TestLibrary.different_language_proj");
+
+            File.Move(oldProjectFile, newProjectFile);
+
+            var restoreCommand = new RestoreCommand(Log, libraryProjectDirectory, "TestLibrary.different_language_proj");
+
+            restoreCommand
+                .Execute()
+                .Should()
+                .Pass();
+        }
+
+        [Theory]
+        [InlineData("Debug", "DEBUG")]
+        [InlineData("Release", "RELEASE")]
+        [InlineData("CustomConfiguration", "CUSTOMCONFIGURATION")]
+        [InlineData("Debug-NetCore", "DEBUG_NETCORE")]
+        public void It_implicitly_defines_compilation_constants_for_the_configuration(string configuration, string expectedDefine)
+        {
+            var testAsset = _testAssetsManager
+                .CopyTestAsset("AppWithLibraryFS", "ImplicitConfigurationConstants", configuration)
+                .WithSource()
+                .Restore(Log, relativePath: "TestLibrary");
+
+            var libraryProjectDirectory = Path.Combine(testAsset.TestRoot, "TestLibrary");
+
+            var getValuesCommand = new GetValuesCommand(Log, libraryProjectDirectory,
+                "netstandard1.6", "DefineConstants");
+
+            getValuesCommand.ShouldCompile = true;
+            getValuesCommand.Configuration = configuration;
+
+            getValuesCommand
+                .Execute("/p:Configuration=" + configuration)
+                .Should()
+                .Pass();
+
+            var definedConstants = getValuesCommand.GetValues();
+
+            definedConstants.Should().BeEquivalentTo(new[] { expectedDefine, "TRACE", "NETSTANDARD", "NETSTANDARD1_6" });
+        }
+
+        [Theory]
+        [InlineData("netstandard1.6", new[] { "NETSTANDARD", "NETSTANDARD1_6" }, false)]
+        [InlineData("net45", new[] { "NETFRAMEWORK", "NET45" }, true)]
+        [InlineData("net461", new[] { "NETFRAMEWORK", "NET461" }, true)]
+        [InlineData("netcoreapp2.0", new[] { "NETCOREAPP", "NETCOREAPP2_0" }, false)]
+        public void It_implicitly_defines_compilation_constants_for_the_target_framework(string targetFramework, string[] expectedDefines, bool buildOnlyOnWindows)
+        {
+            bool shouldCompile = true;
+
+            var testAsset = _testAssetsManager
+                .CopyTestAsset("AppWithLibraryFS", "ImplicitFrameworkConstants", targetFramework)
```

Comment:
The second arg here, if not defaulted to CallerMemberName needs to be unique, but I think this is copied from another test. Consider just making this: `identifier: targetFramework`.